cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/classloader/SparkClassRewriter.java

```diff
@@ -663,6 +669,108 @@ public void visitMethodInsn(int opcode, String owner, String name, String desc,
     return cw.toByteArray();
   }
 
+  /**
+   * Rewrites the temporary file name used for checkpoints from checkpointDir/temp to
+   * checkpointDir/temp-(current_timestamp_milliseconds).
+   *
+   * @param byteCodeStream {@link InputStream} for reading in the original bytecode.
+   * @return the rewritten bytecode
+   */
+  private byte[] rewriteTempFileNameForCheckpoint(InputStream byteCodeStream) throws IOException {
+    String checkpointTimeFieldName =
+      "org$apache$spark$streaming$CheckpointWriter$CheckpointWriteHandler$$checkpointTime";
+    Type timeType = Type.getType(org.apache.spark.streaming.Time.class);
+    Type hadoopPathType = Type.getType(org.apache.hadoop.fs.Path.class);
+    Type stringType = Type.getType(String.class);
+    Type stringBuilderType = Type.getType(StringBuilder.class);
+
+    ClassReader cr = new ClassReader(byteCodeStream);
+    ClassWriter cw = new ClassWriter(0);
+
+    cr.accept(new ClassVisitor(Opcodes.ASM5, cw) {
+      @Override
+      public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
+
+        MethodVisitor mv = super.visitMethod(access, name, desc, signature, exceptions);
+        final GeneratorAdapter adapter = new GeneratorAdapter(mv, access, name, desc);
+        final AtomicBoolean tempStringAddedToStack = new AtomicBoolean(false);
```

Comment:
Move this as a simple `boolean` field in the returned inner class.

Reply:
Done.